pkg/ibmcloudprovider/volume_provider.go

```diff
@@ -138,21 +143,45 @@ func (icp *IBMCloudStorageProvider) GetProviderSession(ctx context.Context, logg
 			PassthroughSecret: string([]byte{}), // // TODO~ Need to remove it
 		}
 	}
+
+	prov, err := icp.Registry.Get(icp.ProviderName)
+	if err != nil {
+		logger.Error("Not able to get the said provider, might be its not registered", local.ZapError(err))
+		return nil, err
+	}
+
 	vpcBlockConfig := &vpcconfig.VPCBlockConfig{
 		VPCConfig:    icp.ProviderConfig.VPC,
 		IKSConfig:    icp.ProviderConfig.IKS,
 		APIConfig:    icp.ProviderConfig.API,
 		ServerConfig: icp.ProviderConfig.Server,
 	}
-	session, isFatal, err := provider_util.OpenProviderSessionWithContext(ctx, vpcBlockConfig, icp.Registry, icp.ProviderName, logger)
-	if err != nil || isFatal {
+
+	for retryCount := 0; retryCount < utils.MaxRetryAttemptForSessions; retryCount++ {
+		session, _, err := provider_util.OpenProviderSessionWithContext(ctx, prov, vpcBlockConfig, icp.ProviderName, logger)
+		if err == nil {
+			logger.Info("Successfully got the provider session", zap.Reflect("ProviderName", session.ProviderName()))
+			return session, nil
+		}
 		logger.Error("Failed to get provider session", zap.Reflect("Error", err))
+		if retryCount == 1 {
+			return nil, err
+		}
+		if providerError, ok := err.(provider.Error); ok && providerError.Code() == reasoncode.ErrorFailedTokenExchange && (strings.Contains(strings.ToLower(providerError.Error()), "api key could not be found") || strings.Contains(strings.ToLower(providerError.Error()), "user not found or active")) {
```

Comment:
can we keep these hardcoded strings in some const file or variable

Reply:
Done